server/modules/doppler-client.js

```diff
@@ -0,0 +1,145 @@
+const querystring = require('querystring');
+const fetch = require('node-fetch');
+const shopify = require ('./shopify-extras');
+const baseUrl = 'https://restapi.fromdoppler.com';
+
+const getCustomerFieldValue = function(customer, fieldPath) {
+    let currentProperty = customer;
+    
+    fieldPath.split(".").forEach(propertyName => {
+        currentProperty = currentProperty[propertyName];
+    });
+    
+    return currentProperty;
+}
+
+class Doppler {
+    constructor(accountName, apiKey) {
+        this.accountName = querystring.escape(accountName);
+        this.apiKey = querystring.escape(apiKey);
+    }
+
+    async sendRequestAsync(url, fetchOptions) {
+        const response = await fetch(url, fetchOptions);
+        const responseBody = await response.json();
+
+        if (response.status >= 400) {
+            const msg = responseBody && (responseBody.title || responseBody.detail)
+                ? `${responseBody.title ? `${responseBody.title}: ` : ''}${responseBody.detail ? responseBody.detail : ''}`
+                : 'Unexpected error';
+    
+            throw new Error(msg);
+        }
+
+        return responseBody
+    }
+
+    async AreCredentialsValidAsync() {
+        const url = `${baseUrl}/accounts/${this.accountName}?api_key=${this.apiKey}`;
+        
+        try {
+            await this.sendRequestAsync(url);
+            return true;
+        } catch (error) {
+            console.warn(`Error validating credentials: ${JSON.stringify(error)}`);
+            return false;
+        }
+    }
+
+    async getListsAsync() {
+        // TODO: what if there are more than 200 lists?
+        const url = `${baseUrl}/accounts/${this.accountName}/lists?page=1&per_page=200&state=active&api_key=${this.apiKey}`;
+
+        const responseBody = await this.sendRequestAsync(url);
+
+        return responseBody.items.map(list => { return { listId: list.listId, name: list.name } });
+    }
+
+    async createListAsync(listName) {
+        const url = `${baseUrl}/accounts/${this.accountName}/lists?api_key=${this.apiKey}`;
+
+        const responseBody = await this.sendRequestAsync(url,  { method:'POST', body: JSON.stringify({name: listName}) });
+        
+        return responseBody.createdResourceId;
+    }
+
+    async getFieldsAsync() {
+        const url = `${baseUrl}/accounts/${this.accountName}/fields?api_key=${this.apiKey}`;
```

Comment:
What do you think about using a library to build URIs? 

I think that it will be useful in some edge cases, for example when the _accountName_ has special characters like `ppascual+test@gmail.com`, if I am not wrong, the symbol `+` should be encoded, in another way it will be parsed as a space.

Reply:
We're escaping the account name, in the constructor of the class, we use the querystring.escape() function to escape both: accountName and apiKey.

Reply:
Ok, a URI builder could help in a more general way, but in that case, I think that it is fine.